contracts/router/tree/TokenTree.sol

```diff
@@ -0,0 +1,578 @@
+    // - The highest byte is always the node index. This is always i, but we store this for consistency.
+    // - The remaining bytes are indexes of the nodes on the path from the node to the root (from highest to lowest).
+    // This way the finding the lowest common ancestor of two nodes is reduced to finding the lowest differing byte
+    // in node's encoded root paths.
+    uint256[] internal _rootPath;
+
+    // (node => bitmask with all attached pools to the node).
+    // Note: This excludes the pool that connects the node to its parent.
+    mapping(uint256 => uint256) internal _attachedPools;
+
+    // ════════════════════════════════════════════════ CONSTRUCTOR ════════════════════════════════════════════════════
+
+    constructor(address bridgeToken) {
+        // The root node is always the bridge token
+        _addNode({token: bridgeToken, depth: 0, poolIndex: 0, rootPathParent: 0});
+        // Push the empty pool so that `poolIndex` for non-root nodes is never 0
+        _pools.push(address(0));
+    }
+
+    // ══════════════════════════════════════════════ INTERNAL LOGIC ═══════════════════════════════════════════════════
+
+    /// @dev Adds a pool with `N = tokensAmount` tokens to the tree by adding N-1 new nodes
+    /// as the children of the given node. Given node needs to represent a token from the pool.
+    function _addPool(
+        uint256 nodeIndex,
+        address pool,
+        address poolModule,
+        uint256 tokensAmount
+    ) internal {
+        require(nodeIndex < _nodes.length, "Out of range");
+        Node memory node = _nodes[nodeIndex];
+        if (poolModule == address(0)) poolModule = address(this);
+        (bool wasAdded, uint8 poolIndex) = (false, _poolMap[pool].index);
+        if (poolIndex == 0) {
+            poolIndex = uint8(_pools.length);
+            require(poolIndex <= type(uint8).max, "Too many pools");
+            _pools.push(pool);
+            _poolMap[pool] = Pool({module: poolModule, index: poolIndex});
+            wasAdded = true;
+        } else {
+            // Check if the existing pool could be added to the node. This enforces some sanity checks,
+            // as well the invariant that any path from root to node doesn't contain the same pool twice.
+            _checkPoolAddition(nodeIndex, node.depth, poolIndex);
+        }
+        // Remember that the pool is attached to the node
+        _attachedPools[nodeIndex] |= 1 << poolIndex;
+        address[] memory tokens = _getPoolTokens(poolModule, pool, tokensAmount);
+        bool nodeFound = false;
+        unchecked {
+            uint8 childDepth = node.depth + 1;
+            uint256 rootPathParent = _rootPath[nodeIndex];
+            for (uint256 i = 0; i < tokensAmount; ++i) {
+                address token = tokens[i];
+                // Save token indexes if this is a new pool
+                if (wasAdded) {
+                    tokenIndexes[pool][token] = uint8(i);
```

Comment:
Similar unsafe cast to uint8 issues here given `_getPoolTokens` from an external pool module doesn't necessarily have to be restricted to length <= `type(uint8).max` (see e.g. Curve v1 as `int128` for `N_COINS`)

Reply:
Token Tree supports up to 256 nodes, including the root. So adding a pool that has indexes that don't fit into `uint8` will not be possible due to check in `_addNode()`. Therefore, if a pool gets added, we could safely assume its token indexes will fit into `uint8`, allowing the unsafe cast. Will try to reflect that in the comments.

Reply:
To be sure I understand correctly: we always have `tokens.length` <= `_nodes.length`. So if `numTokens > type(uint8).max`, we'd also have `_nodes.length > type(uint8).max`, which would cause a revert on `_addNode()`?

Reply:
This is 100% correct!